compiler/sem/semexprs.nim

```diff
@@ -696,85 +696,135 @@ proc arrayConstrType(c: PContext, n: PNode): PType =
+    # an empty array constructor
+    let indexType = getSysType(c.graph, n.info, tyInt)
+    rawAddSon(result.typ, createRange(Zero, toInt128(-1), indexType))
     rawAddSon(result.typ, newTypeS(tyEmpty, c)) # needs an empty basetype!
-    lastIndex = toInt128(-1)
   else:
-    var x = n[0]
-    if x.kind == nkExprColonExpr and x.len == 2:
-      var idx = semConstExpr(c, x[0])
-      if not isOrdinalType(idx.typ):
-        result = n
-        result[0][0] = c.config.newError(x[0],
-                          PAstDiag(kind: adSemExpectedOrdinalArrayIdx,
-                                   nonOrdInput: idx,
-                                   indexExpr: result))
-        # TODO: the error message references the index expression (`result`)
-        #       and we're wrapping the error anyway, maybe expand the nkError
-        #       to cover the whole result in one go
-        result = c.config.wrapError(result)
-        return
-      else:
-        firstIndex = getOrdValue(idx)
-        lastIndex = firstIndex
-        indexType = idx.typ
-        lastValidIndex = lastOrd(c.config, indexType)
-        x = x[1]
-
-    let yy = semExprWithType(c, x)
-    var typ = yy.typ
-    result.add yy
-    for i in 1..<n.len:
-      if lastIndex == lastValidIndex:
-        let validIndex = makeRangeType(
-          c, toInt64(firstIndex),
-          toInt64(lastValidIndex), n.info, indexType)
-
+    # analyse the first element separately -- the other elements depend on its
+    # type
+    var
+      (first, firstIndex, indexType) = semArrayElement(c, n[0], nil)
+      lastIndex = firstIndex ## tracks the index of the previous element
+
+    if indexType == nil:
+      # if no type is provided or an error occurred, default to ``int`` for
+      # the index type
+      indexType = getSysType(c.graph, n.info, tyInt)
+
+    # before doing anything else, check if the size of the array doesn't exceed
+    # the maximum value representable by the index type
+    let limit = lastOrd(c.config, indexType)
+    if firstIndex + toInt128(n.len - 1) > limit:
         result = c.config.newError(n,
                             PAstDiag(
                               kind: adSemIndexOutOfBounds,
-                              ordRange: validIndex,
-                              maxOrdIdx: i,
-                              outOfBoundsIdx: n.len))
+                              ordRange: createRange(firstIndex, limit, indexType),
+                              outOfBoundsIdx: toInt(firstIndex) + n.len - 1))
         return
 
-      x = n[i]
-      if x.kind == nkExprColonExpr and x.len == 2:
-        var idx = semConstExpr(c, x[0])
-        idx = fitNode(c, indexType, idx, x.info)
-        if lastIndex + 1 != getOrdValue(idx):
-          result = n
-          result[i] = c.config.newError(x, 
-                          PAstDiag(kind: adSemInvalidOrderInArrayConstructor))
-          result = c.config.wrapError(result)
-          return
+    if first.kind != nkError:
```

Comment:
With the `inferredIndexType` as a separate definition the condition `first.kind != nkError and inferredIndexType.isNil` would let you choose the starting index for the for loop below (currently hardcode to `1`). Doing that would mean all elements are handled the same way and a slight reduction in duplicated logic.

Reply:
I'm not sure I understand. In the case that `first.kind != nkError and inferredIndexType != nil`, separate handling is still required, otherwise the first element would not be analysed or added.

Reply:
I started writing it out and it won't simplify without reworking `semArrayElement` to be more consistent in its returned element. I only suggested on consistency scheme there, but others would also work, such as: returning two `PNode` one for the index and the other for the element, if only the element is present the index is empty; or whatever else.

It's not a must, only if you want to. I don't think it necessarily will be more efficient, it might execute extra analysis when errors have already been found, it would simply read more linearly.

Reply:
Done. `semArrayElement` is renamed to `semArrayElementIndex` and now does what you described.